### PR TITLE
[#8527] Migrate to JUnit 5 Assertions API

### LINKED ISSUE #8527: Explore migrating to JUnit 5
At one point we decided to move from JUnit to TestNg due to the higher flexibility of test configuration allowed by TestNG. It may be good to revisit this decision to see if it is still valid and if not, to migrate back to JUnit.
Reasons: 
* JUnit is the more popular of the two
* JUnit is faster to keep pace with Java's evolution while TestNG doesn't even seem to be evolving any more.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Part of the original #8527.

**Outline of Solution**

Used JUnit 5 Assertions API. Most of the current `assert*` usages are unaffected, but we now have a new (strong) member: `assertThrows`.

This construct:
```java
try {
    doSomethingThatWillThrowException();
    failAsThisShouldNotBeReachable();
} catch (SomeException e) {
    processException(e);
}

// or

@Test(expectedExceptions = SomeException.class)
public void testSomething() {
    doSomethingThatWillThrowException();
}
```

can be transformed to:

```java
SomeException e = assertThrows(SomeException.class, () -> doSomethingThatWillThrowException());
processException(e);
```

which is the best of both worlds.

Note that action tests are not migrated to this new technique because they will be affected by V7 migration anyway.

### CODE REVIEW

File: src/test/java/teammates/test/cases/logic/StudentsLogicTest.java [hunk 1]
```diff
@@ -261,28 +261,30 @@ private void testValidateSections() throws Exception {
...
                 .withComments("")
                 .build());
-        try {
-            studentsLogic.validateSectionsAndTeams(studentList, courseId);
-        } catch (EnrollException e) {
-            assertEquals(String.format(Const.StatusMessages.TEAM_INVALID_SECTION_EDIT, "Team 1.1</td></div>'\"")
-                             + "Please use the enroll page to edit multiple students",
-                         e.getMessage());
-        }
+        studentList.add(StudentAttributes
```
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/test/java/teammates/test/cases/logic/StudentsLogicTest.java]: The previous one was missing the `signalFailureToDetectException()` and the exception was never actually thrown.

File: src/test/java/teammates/test/cases/action/StudentFeedbackResultsPageActionTest.java [hunk 2]
```diff
@@ -185,7 +187,7 @@ public void testExecuteAndPostProcess() throws Exception {
         pageAction = getAction(submissionParams);
 
         try {
-            pageResult = getShowPageResult(pageAction);
+            getShowPageResult(pageAction);
```
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/test/java/teammates/test/cases/action/StudentFeedbackResultsPageActionTest.java]: This is another one of the test cases where the exception is never actually thrown. Reason being the expected exception is that the session is closed but not yet visible, but the session is already closed and published. I tried modifying the test data but it has quite a widespread effect.
I wanted to actually just remove this test considering all action tests will be affected anyway in V7, do you think it's a wise decision?

File: static-analysis/teammates-macker.xml [hunk 3]
```diff
@@ -295,6 +295,7 @@
                         <allow>
                             <from>
                                 <include class="${testCases}.datatransfer.**" />
+                                <include class="${testCases}.storage.**" />
```
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply static-analysis/teammates-macker.xml]: There actually is that kind of reference, for example:
```java
try {
    someDb.createEntity(...);
    signalFailureToDetectException();
} catch (SomeException e) {
    ...
}
```
`createEntity` obviously creates a reference to the entity object. It looks like converting it to a lambda body makes that reference explicit.

But this is a correct rule, so there is no big deal. It's just interesting how Macker does not flag it before.

File: src/test/java/teammates/test/cases/action/StudentFeedbackResultsPageActionTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (xpdavid) [reply src/test/java/teammates/test/cases/action/StudentFeedbackResultsPageActionTest.java]: I think we should not remove it. When writing test cases for V7, I guess most likely we will use existing test cases as references. This test scenario is valid and might even be important (accessing closed session which is not published).

I think we can have a little "hack" to make it work without modifying test data.

```java
        Instant backupResultVisibleTime = closedSession.getResultsVisibleFromTime();
        // session is closed but not published
        closedSession.setResultsVisibleFromTime(Instant.now().plus(Duration.ofDays(1L)));
        FeedbackSessionsLogic.inst().updateFeedbackSession(closedSession);

        // .....

        closedSession.setResultsVisibleFromTime(backupResultVisibleTime);
        FeedbackSessionsLogic.inst().updateFeedbackSession(closedSession);
```

File: src/test/java/teammates/test/cases/datatransfer/StudentAttributesFactoryTest.java [hunk 4]
```diff
@@ -15,110 +15,102 @@
...
-        try {
-            saf.makeStudent(line, courseId);
-            signalFailureToDetectException();
-        } catch (EnrollException e) {
-            assertEquals(StudentAttributesFactory.ERROR_ENROLL_LINE_EMPTY, e.getMessage());
-        }
+        StudentAttributesFactory saf = new StudentAttributesFactory();
+        String line = "";
+        String courseId = "SAFT.courseId";
+        EnrollException ee = assertThrows(EnrollException.class, () -> saf.makeStudent(line, courseId));
```
Comment (damithc) [inline src/test/java/teammates/test/cases/datatransfer/StudentAttributesFactoryTest.java]: Ensure line 91 is executed. In the previous versions of JUnit, anything below the exception-causing line is ignored silently.

File: src/main/java/teammates/common/util/Assumption.java [hunk 5]
```diff
@@ -8,9 +8,7 @@
  * an unchecked {@link AssertionError} will be thrown at runtime.
  *
  * <p>Normally, we use Java assertion to do runtime checking, but GAE does not support the assertions.
- * The methods of this file is adapted from org.junit.Assert v4.10.
- *
- * @see org.junit.Assert
+ * The methods of this file are adapted from org.junit.jupiter.api.Assertions v5.3.1.
```
Comment (damithc) [inline src/main/java/teammates/common/util/Assumption.java]: My comment is not related to this PR: The reason we introduced the Assumptions class was because earlier versions of GAE did not support the native assert statement. May want to check if we still need this class.

File: src/test/java/teammates/test/cases/BaseTestCase.java [hunk 6]
```diff
@@ -124,87 +115,79 @@ protected static String getPopulatedEmptyStringErrorMessage(String messageTempla
      */
 
     protected static void assertTrue(boolean condition) {
-        AssertJUnit.assertTrue(condition);
+        Assertions.assertTrue(condition);
```
Comment (damithc) [inline src/test/java/teammates/test/cases/BaseTestCase.java]: This comment is not related to this PR: use a static import so that `Assertions.` part can be omitted?

Comment (wkurniawan07) [reply src/test/java/teammates/test/cases/BaseTestCase.java]: Not possible; immediate name clash.

File: src/test/java/teammates/test/cases/datatransfer/StudentAttributesFactoryTest.java (hunk 4 above)
Comment (wkurniawan07) [reply src/test/java/teammates/test/cases/datatransfer/StudentAttributesFactoryTest.java]: That's older JUnit which we're moving on from. `assertThrows` has the ability of capturing (only) the desired exception without breaking the program flow.
